Editor: Performance Improvements for wplink

timmyc opened the pull request:
This branch seeks to close #12481 by implementing a few performance enhancements to the `wplink` TinyMCE plugin.  The original issue was reported around 3 months ago in the forums by a user who has a site with over 1600 posts.

The user supplied a video showing the issue in action, which involves opening the link dialog multiple times in the same post, and how over time, the action continually gets slower.  While digging in to the problem I noticed that each time the link dialog is opened, we request the first page of posts for the site to populate the recent posts list.  This request can be quite slow on sites with a large number of posts.

Additionally, we do a request for `postTypes` each time the dialog is opened.

__To Reproduce__
It is best to play around on a site with a large number of posts to experience the sluggish behavior.  I used en.blog myself, and sandboxing the API really shows the slow behavior further.  Just open up the editor, and click the link button in the MCE toolbar multiple times.  Note that you can't really even cancel out of the dialog by clicking on the "Cancel" button immediately after it opens.

__Proposed Solution__
In this branch I've added logic to track if the wplink dialog has been shown once already, and as such the first page of posts have been fetched once already.  On subsequent displays of the dialog, a prop is passed to `<PostSelector />` to `surpressFirstPageLoad` if the API request has already taken place, and `posts` have been returned from the state tree for the first page of posts.

Also, the `QueryPostTypes` call is only done if no `postTypes` are passed in as a prop.  Even though the `postTypes` API call is a fast one, I figure it is worth preventing un-needed calls.

I also added logic to `getQuery` in the `PostSelector` to order search queries by `date` in an effort to mimic the search results seen in wp-admin.  I haven't been able to get the results to fully match what happens in wp-admin, but I think that is due to the API passing in more arguments to `WP_Query` - but I didn't fully go down that 🐰 🕳 

The last change I am proposing here is increasing the debounce time on the search field in the `<PostSelector />` by 300ms.  Just to _filter_ out potential API requests if someone types in the search field slower.

__To Test__
- Open the editor for a site with a large number of posts
- Have your network tab open in your browser's console
- Open the `wplink` dialog, note an API request is sent to get a list of posts for the site
- clear out your network tab, close the dialog, and open it again.  Note that no API requests should be sent for posts or postTypes
- Perform some searches in the dialog and verify that still works as expected

supernovia commented:
I did a bit of testing; I was able to add multiple links in an article without much trouble. I did notice though that if I scrolled things bogged down again, FWIW.

aduth commented:
>The user supplied a video showing the issue in action, which involves opening the link dialog multiple times in the same post, and how over time, the action continually gets slower. 

Makes me wonder if this is really an issue with response time, or if we're accumulating garbage every time we're opening the modal. In you're reproducing, did you give the response enough to finish before closing and reopening?

>we request the first page of posts for the site to populate the recent posts list. This request can be quite slow on sites with a large number of posts.

Ideally the response time should not scale proportionally to the number of posts. Are there optimizations we can make server-side?

lancewillett commented:
Re-ping @timmyc — in case you missed comments and questions above.

timmyc commented:
Thanks for the ping @lancewillett - I will give it a rebase.

> Makes me wonder if this is really an issue with response time, or if we're accumulating garbage every time we're opening the modal. In you're reproducing, did you give the response enough to finish before closing and reopening?

@aduth all valid and great questions!  I do think there is a deeper issue here, perhaps one related to `react-virtualized`, and most **definitely** there could be some tuning done to the posts endpoint the help speed up response times on sites with large numbers of posts.  In this branch I'm just trying to help alleviate what I think are un-necessary fetches of the same first page of posts from the API when using `wplink`.

As far as my usage to trigger the slow down... I tried to emulate the steps provided by the original bug reporter.  Open up the link dialog, search for a post, select it, and close the dialog.  The opening of the dialog would trigger the first page load each time, and then using the search would then fire off the filtered query.  Sometimes the search query would return prior to the 1st page.  Although not the ultimate performance fix for this flow, not requesting the first page on each open of wplink, and increasing the debounce time on search did improve performance in my testing.

timmyc commented:
@aduth wondering if you might be able to give this one last quick look to see if we could merge it in.  @supernovia was inquiring about the issue again so would like to move forward with this if we could.  Rebased, fixed linting errors.

aduth commented on client/my-sites/post-selector/index.jsx:
```diff
@@ -63,13 +65,22 @@ export default React.createClass( {
 			}
 
+			// if we don't have a search term, default to ordering by date
+			if ( key === 'orderBy' && search !== '' ) {
```
Aren't these the default anyways?

aduth commented on client/my-sites/post-selector/selector.jsx:
```diff
@@ -393,7 +393,7 @@ const PostSelectorPosts = React.createClass( {
 
 	render() {
-		const { className, siteId, query } = this.props;
+		const { className, siteId, query, surpressFirstPageLoad, posts, postTypes } = this.props;
```
Typo: `surpress` -> `suppress`

aduth commented on client/my-sites/post-selector/selector.jsx:
```diff
@@ -403,15 +403,22 @@ const PostSelectorPosts = React.createClass( {
 			'is-type-labels-visible': isTypeLabelsVisible
 		} );
 
+		const pagesToRequest = filter( requestedPages, ( page ) => {
```
Minor: And uncertain how well it plays with `key`, but instead of iterating `requestedPages` twice (`filter`, then `map`), could the `map` return value return `null` for the first page if we're suppressing?
